src/iterate.jl

```diff
@@ -0,0 +1,138 @@
+
+export knots
+
+# Similar to ExtrapDimSpec but for only a single dimension
+const ExtrapSpec = Union{BoundaryCondition,Tuple{BoundaryCondition,BoundaryCondition}}
+
+# Macro to get create ExtrapSpec for checking if a KnotIterator has a given BC
+# for forward iteration
+macro FwdExtrapSpec(bc)
+    :( Union{$bc,Tuple{BoundaryCondition,$bc}} )
+end
+struct KnotIterator{T,ET <: ExtrapSpec}
+    knots::Vector{T}
+    bc::ET
+    nknots::Int
+    KnotIterator{T,ET}(k::AbstractArray{T}, bc::ET) where {T,ET} = new(k, bc, length(k))
+end
+
+KnotIterator(k::NTuple{N,AbstractArray}, bc::NTuple{N, ExtrapSpec}) where {N} = map(KnotIterator, k, bc)
+KnotIterator(k::NTuple{N,AbstractArray}, bc::BoundaryCondition) where {N} = map(x -> KnotIterator(x, bc), k)
+KnotIterator(k::AbstractArray{T}, bc::ET) where {T,ET <: ExtrapDimSpec} = KnotIterator{T,ET}(k, bc)
+
+const RepeatKnots = Union{Periodic,Reflect}
+Base.IteratorSize(::Type{KnotIterator}) = Base.SizeUnknown()
```

Comment:
Rather than prefixing `Base` like this, perhaps we should explicitly import these in Interpolations.jl ?